component/kafka/simple/component.go

```diff
@@ -0,0 +1,361 @@
+package simple
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/beatlabs/patron/component/kafka"
+	patronErrors "github.com/beatlabs/patron/errors"
+	"github.com/beatlabs/patron/internal/validation"
+	"github.com/beatlabs/patron/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	consumerComponent = "kafka-consumer"
+	subsystem         = "kafka_simple"
+	messageReceived   = "received"
+	messageProcessed  = "processed"
+	messageErrored    = "errored"
+	messageSkipped    = "skipped"
+)
+
+const (
+	defaultRetries         = 3
+	defaultRetryWait       = 10 * time.Second
+	defaultBatchSize       = 1
+	defaultBatchTimeout    = 100 * time.Millisecond
+	defaultFailureStrategy = kafka.ExitStrategy
+)
+
+var (
+	consumerErrors           *prometheus.CounterVec
+	topicPartitionOffsetDiff *prometheus.GaugeVec
+	messageStatus            *prometheus.CounterVec
+)
+
+func init() {
+	consumerErrors = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "component",
+			Subsystem: subsystem,
+			Name:      "consumer_errors",
+			Help:      "Consumer errors, classified by consumer name",
+		},
+		[]string{"name"},
+	)
+
+	topicPartitionOffsetDiff = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "component",
+			Subsystem: subsystem,
+			Name:      "offset_diff",
+			Help:      "Message offset difference with high watermark, classified by topic and partition",
+		},
+		[]string{"group", "topic", "partition"},
+	)
+
+	messageStatus = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "component",
+			Subsystem: subsystem,
+			Name:      "message_status",
+			Help:      "Message status counter (received, processed, errored) classified by topic and partition",
+		}, []string{"status", "partition", "topic"},
+	)
+
+	prometheus.MustRegister(consumerErrors, topicPartitionOffsetDiff, messageStatus)
+}
+
+type Component struct {
+	name                     string
+	topic                    string
+	brokers                  []string
+	saramaConfig             *sarama.Config
+	proc                     kafka.BatchProcessorFunc
+	failStrategy             kafka.FailStrategy
+	batchSize                uint
+	batchTimeout             time.Duration
+	retries                  uint
+	retryWait                time.Duration
+	partitionConsumerFactory func(ctx context.Context) (sarama.Client, sarama.Consumer, map[int32]sarama.PartitionConsumer, error)
+	// DurationOffset options
+	durationBasedConsumer bool
+	durationOffset        time.Duration
+	timeExtractor         func(*sarama.ConsumerMessage) (time.Time, error)
+	// NotificationOnceReachingLatestOffset options
+	latestOffsetReachedChan chan<- struct{}
+	startingOffsets         map[int32]int64
+	latestOffsets           map[int32]int64
+	once                    sync.Once
+}
+
+// New initializes a new simple kafka consumer component with support for functional configuration.
```

Comment:
"simple" is perhaps the most overrated terms in the history of software engineering (and also subjective, varying from person to person); do you think we can qualify better what this consumer does vs existing ones? the guarantees and qualities for example.